### ISSUE TITLE
Change setting redirect link from Scratch to our domain

### ISSUE THREAD

**Issue by Hans5958** [ScratchAddons/ScratchAddons#558]
I'm not sure why we have https://scratch.mit.edu/scratch-addons-extension/settings for linking the addon settings page, while we can do something in our domain like https://scratchaddons.com/settings.

**Comment by JeffaloBob** [comment]
we don't have access to scratchaddons, and a change like that might disable the addon for everyone until they accept

**Comment by WorldLanguages** [comment]
> we don't have access to scratchaddons.com, and a change like that might disable the addon for everyone until they accept

Yeah, we don't want to request any more host permissions, and we can only read the URL from tabs where we have host permissions. Websites can't open `chrome-extension://` URLs directly, they require help from an extension.

**Comment by Hans5958** [comment]
I see. Wouldn't it be shorter if it's like https://scratch.mit.edu/#sa-settings, or https://scratch.mit.edu/#scratch-addons-settings (so it won't link a 404 page), or https://scratch.mit.edu/scratch-addons/settings?

**Comment by WorldLanguages** [comment]
Well you see, if it links to a 404 it will likely load faster. And yeah it could be shorter, I just preferred to make it longer since we're basically using Scratch's domain for our own purposes so better safe than sorry.

**Comment by Hans5958** [comment]
> if it links to a 404 it will likely load faster.

I don't think so. I got redirected way before it loads anything. It is instantaneous, so it doesn't matter where we pointing to.

**Comment by WorldLanguages** [comment]
Yeah, actually, the scratchr2 page takes more to load than the homepage, since it's SSR vs react based. I think that in order to answer which one is actually faster we'd need to know more about how browsers work tho lol, I won't be looking into that.

**Comment by Explosion-Scratch** [comment]
That's cool. I didn't know we had that loll

**Comment by WorldLanguages** [comment]
@Hans5958 You should reconsider to close this, after what I said on that issue.

**Comment by Hans5958** [comment]
I'll keep this open for future possibilities.

**Comment by mxmou** [comment]
> I'll keep this open for future possibilities.

I don't think there are any possibilities for this other than adding an unnecessary permission.

**Comment by Hans5958** [comment]
You are not a fortune-teller, and I'm also not.

**Comment by Hans5958** [comment]
Very overdue update, but the only certain opportunity of the domain gets added is when https://scratch.org is needed to be added. Not sure if that will happen sometime in the future, or when, but it's good to know that there is one.

https://discord.com/channels/751206349614088204/751207339545198682/801295201757954058

Maybe this also can be included on manifest v3 updates? I'm not sure if updating from a v2 extension to a v3 extension will require user to agree with the update.

**Comment by WorldLanguages** [comment]
> I'm not sure if updating from a v2 extension to a v3 extension will require users to agree with the permissions.

Of course. If that's not the cause, that's likely worth reporting in the Chrome bug bounty program...

**Comment by Hans5958** [comment]
> > I'm not sure if updating from a v2 extension to a v3 extension will require users to agree with the permissions.
> 
> Of course. If that's not the cause, that's likely worth reporting in the Chrome bug bounty program...

So you can add it with the manifest v3 update?

**Comment by WorldLanguages** [comment]
@Hans5958 No? I meant the opposite. Manifest version shouldn't be affecting how Chrome reacts to permission changes.

**Comment by Hans5958** [comment]
Ah, your wording seemed weird.

**Comment by Hans5958** [comment]
Next chance is #2599?

**Comment by WorldLanguages** [comment]
Not really, since min browser version is unrelated to permissions
Real chance will be when Chrome removes support for required host permissions (we'd need to request them as optional permissions after install). That will surely happen before Scratch migrates to Scratch.org, if ever.
Whenever that happens, we can ask scratchaddons.com permission for new users.
Old users would be stuck without that permission unless we force them to accept it (which is not reasonable for a simple redirect imo)
If Scratch ever migrates to Scratch.org, which will require both old and new users to accept new optional permissions, we can add `*.scratch.org` and `*.scratchaddons.com`  there.

**Comment by WorldLanguages** [comment]
@Hans5958 
1. It's `delarativeNetRequest`, not `delarativeWebRequest`
2. `declarativeNetRequest` does not need any permission prompts by design

**Comment by Hans5958** [comment]
> `declarativeNetRequest` does not need any permission prompts by design

How certain are you if that's the case?

**Comment by mxmou** [comment]
> `declarativeNetRequest` does not need any permission prompts by design

From Chrome documentation:
![image](https://user-images.githubusercontent.com/51849865/119867818-997d9380-bf1e-11eb-8570-7dd6d87a20a8.png)

**Comment by WorldLanguages** [comment]
Bruh
Literally me rn: :|
Why would they do that
Why would a less powerful API than webRequest have a warning but not webRequest
Doesn't make any sense

**Comment by Hans5958** [comment]
Well, it seems that uses `declarativeNetRequest` is going to result the prompt after all.

### Version 1 manifest (Manifest version 2)

```json
{
  "name": "Version 1",
  "version": "0.0.1",
  "manifest_version": 2,
  "description": "This is a test extension.",
  "homepage_url": "https://hans5958.github.io",
  "icons": {
    "16": "icons/icon16.png",
    "48": "icons/icon48.png",
    "128": "icons/icon128.png"
  },
  "permissions": [
	"https://scratch.mit.edu/*",
    "https://api.scratch.mit.edu/*",
    "https://clouddata.scratch.mit.edu/*",
    "cookies",
    "webRequest",
    "webRequestBlocking",
    "storage",
    "contextMenus",
    "alarms"
  ],
  "default_locale": "en",
  "options_page": "src/options/index.html",
  "browser_action": {
    "default_icon": "icons/icon19.png",
    "default_title": "browser action demo",
    "default_popup": "src/browser_action/browser_action.html"
  }
}
```

### Version 2 manifest (Manifest version 3)

```json
{
  "name": "Version 2",
  "version": "0.0.2",
  "manifest_version": 3,
  "description": "This is a test extension.",
  "homepage_url": "https://hans5958.github.io",
  "icons": {
    "16": "icons/icon16.png",
    "48": "icons/icon48.png",
    "128": "icons/icon128.png"
  },
  "host_permissions": [
    "https://scratch.mit.edu/*",
    "https://api.scratch.mit.edu/*",
    "https://clouddata.scratch.mit.edu/*"
  ],
  "permissions": [
    "cookies",
    "declarativeNetRequest",
    "storage",
    "contextMenus",
    "alarms"
  ],
  "default_locale": "en",
  "options_page": "src/options/index.html",
  "action": {
    "default_icon": "icons/icon19.png",
    "default_title": "browser action demo",
    "default_popup": "src/browser_action/browser_action.html"
  }
}
```

### Images

![image](https://user-images.githubusercontent.com/11584103/119982894-a2876700-bfe9-11eb-9d98-a9048bc757eb.png)
![image](https://user-images.githubusercontent.com/11584103/119983151-fbef9600-bfe9-11eb-8a95-90f4cc688ad0.png)

### Attachment

[poc.zip](https://github.com/ScratchAddons/ScratchAddons/files/6560778/poc.zip)

**Comment by WorldLanguages** [comment]
@Hans5958 Yeah you're right. I sent feedback to Chrome about this. It's stupid.

**Comment by mxmou** [comment]
@WorldLanguages Currently host permissions aren't required to block content using `declarativeNetRequests`, but I'm not sure why. If that's changed or if a separate permission for blocking is added, that warning could be removed.

**Comment by Hans5958** [comment]
> @Hans5958 Yeah you're right. I sent feedback to Chrome about this. It's stupid.

Oh, yeah, on the second image, also pay attention on the first line. The second one is because of `declarariveWhatever`, but the first line is because of moving the URLs to host permissions, maybe?

**Comment by Hans5958** [comment]
So, what's your plan?

**Comment by WorldLanguages** [comment]
This won't happen until at least one of these happens:
1. Chrome removes host permissions being enabled by default when installing
2. Scratch migrates to scratch.org

If number 1 happens, we could request host permissions for `*.scratch.mit.edu` and `*.scratchaddons.com`, for all new users. We could also not disturb old users at all, only disturb them when they interact with the extension, or disturb them all like we did with /scratch-messaging-transition.
If number 2 happens, we could request host permissions for `*.scratch.org` and `*.scratchaddons.com` for all users, and we'll probably have to open a new tab for all of them, because the extension will be completely broken.

`declarativeNetRequest` is not related to this anymore. Only 1 addon and background/handle-cookie.js (for profile comments only) strictly require blocking functionality. So when we migrate to `declarativeNetRequest` it will be an optional permission that's requested when enabling Scratch Notifier.

**Comment by mxmou** [comment]
I'm not sure if I would allow an extension access to its website if it asked me to. The only reason for doing it that I could think of without knowing about this feature is tracking.

**Comment by WorldLanguages** [comment]
Yeah, some users might think that. But reality is that if we wanted to track, we could just set correct CORS headers. No host permissions are needed to ping a server with proper CORS headers.

**Comment by Hans5958** [comment]
> tracking

I mean, considering that...

a. Our audience (mostly) are children and teens who have short attention span (they may don't even know we don't do analytics/tracking, unlike something like that RPC extension)
b. We have privacy statements that might be directed towards, parents, teachers, paranoid people, and CWS
c. There are already ways of tracking that doesn't need host permissons.
d. The extension has links to the site itself.

...I don't see this as a significant issue. But then, I might be wrong. I myself didn't even realised that until you pointed out. So, who knows?

In the other hand, there are some things that we could implement, such as more detailed feedback sending, but then that's for another time. This is just for redirects to the settings page for now.

**Comment by WorldLanguages** [comment]
Note that scratchaddons.com/settings is now redirecting to scratch.mit.edu/scratch-addons-extension/settings after I added a page rule on Cloudflare
